File: python-debugger/app/services/wait_manager.py
from threading import Event, Lock


class WaitManager:
    def __init__(self):
        self._events = {}
        self._lock = Lock()

    def create(self, call_id):
        with self._lock:
            event = self._events.get(call_id)
            if event is None:
                event = Event()
                self._events[call_id] = event
            return event

    def wait(self, call_id, timeout=300):
        event = self.create(call_id)
        ok = event.wait(timeout=timeout)
        with self._lock:
            self._events.pop(call_id, None)
        return "continue" if ok else "timeout_continue"

    def continue_one(self, call_id):
        with self._lock:
            event = self._events.pop(call_id, None)
        if event:
            event.set()
            return True
        return False

    def continue_all(self):
        with self._lock:
            items = list(self._events.values())
            self._events.clear()
        for event in items:
            event.set()
        return len(items)


wait_manager = WaitManager()

File: java-debugger/src/main/java/com/example/microbreakpoint/service/WaitManager.java
package com.example.microbreakpoint.service;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import org.springframework.stereotype.Service;

@Service
public class WaitManager {

    private final ConcurrentHashMap<String, CountDownLatch> waits = new ConcurrentHashMap<>();

    public CountDownLatch create(String callId) {
        return waits.computeIfAbsent(callId, ignored -> new CountDownLatch(1));
    }

    public String waitFor(String callId, long timeoutSeconds) {
        CountDownLatch latch = create(callId);
        boolean released;
        try {
            released = latch.await(timeoutSeconds, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            released = false;
        } finally {
            waits.remove(callId);
        }
        return released ? "continue" : "timeout_continue";
    }

    public boolean continueOne(String callId) {
        CountDownLatch latch = waits.remove(callId);
        if (latch == null) {
            return false;
        }
        latch.countDown();
        return true;
    }

    public int continueAll() {
        List<CountDownLatch> items = new ArrayList<>(waits.values());
        waits.clear();
        items.forEach(CountDownLatch::countDown);
        return items.size();
    }
}
